src/core/utils/validate-ipni-advertisement.ts

```diff
@@ -126,3 +251,262 @@ export async function validateIPNIAdvertisement(
+  const results = response.MultihashResults
+  if (!Array.isArray(results)) {
+    return []
+  }
+
+  return results.flatMap(({ ProviderResults }) => {
+    if (!Array.isArray(ProviderResults)) {
+      return []
+    }
+    return ProviderResults
+  })
+}
+
+/**
+ * Derive expected IPNI multiaddrs from provider information.
+ *
+ * For each provider, attempts to extract their PDP serviceURL and convert it to
+ * the multiaddr format used in IPNI advertisements. This allows validation that
+ * specific providers have advertised the content.
+ *
+ * Note: ProviderInfo should contain the serviceURL at `products.PDP.data.serviceURL`.
+ * In some SDK versions, it may be at the top level. This function checks both locations
+ * to maintain compatibility.
+ *
+ * @param providers - Array of provider info objects from synapse SDK
+ * @param extraMultiaddrs - Additional multiaddrs to include in expectations
+ * @param logger - Optional logger for diagnostics
+ * @returns Expected multiaddrs and count of providers that couldn't be processed
+ */
+function deriveExpectedMultiaddrs(
+  providers: ProviderInfo[],
+  extraMultiaddrs: string[] | undefined,
+  logger: Logger | undefined
+): {
+  expectedMultiaddrs: string[]
+  skippedProviderCount: number
+} {
+  const derivedMultiaddrs: string[] = []
+  let skippedProviderCount = 0
+
+  for (const provider of providers) {
+    // Primary path: products.PDP.data.serviceURL (current SDK structure)
+    // Fallback path: top-level serviceURL (for compatibility with older SDK versions)
+    const serviceURL =
+      provider.products?.PDP?.data?.serviceURL ??
+      (provider as unknown as { serviceURL?: string }).serviceURL ??
+      undefined
+
+    if (!serviceURL) {
+      skippedProviderCount++
+      logger?.warn({ provider }, 'Expected provider is missing a PDP serviceURL; skipping IPNI multiaddr expectation')
+      continue
+    }
+
+    const derivedMultiaddr = serviceURLToMultiaddr(serviceURL, logger)
+    if (!derivedMultiaddr) {
+      skippedProviderCount++
+      logger?.warn({ provider, serviceURL }, 'Unable to derive IPNI multiaddr from serviceURL; skipping expectation')
+      continue
+    }
+
+    derivedMultiaddrs.push(derivedMultiaddr)
+  }
+
+  const additionalMultiaddrs = extraMultiaddrs?.filter((addr) => addr != null && addr !== '') ?? []
+  const expectedMultiaddrs = Array.from(new Set<string>([...additionalMultiaddrs, ...derivedMultiaddrs]))
+
+  return {
+    expectedMultiaddrs,
+    skippedProviderCount,
+  }
+}
+
+/**
+ * Check if any provider in the IPNI response has at least one address.
+ *
+ * This is used for generic IPNI validation when no specific provider is expected.
+ * Passes if IPNI shows ANY provider advertising the content with addresses.
+ *
+ * @param providerResults - Provider results from IPNI response
+ * @returns True if at least one provider has non-empty addresses
+ */
+function hasAnyProviderWithAddresses(providerResults: ProviderResult[]): boolean {
```

Comment:
```suggestion
function hasAnyProviderWithMultiaddrs(providerResults: ProviderResult[]): boolean {
```

Nit: being a little more clear?